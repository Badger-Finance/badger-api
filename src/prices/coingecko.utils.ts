import { InternalServerError } from '@tsed/exceptions';
import { Chain } from '../chains/config/chain.config';
import { COINGECKO_URL } from '../config/constants';
// TODO: generalize and add some axios utilities
import { request } from '../etherscan/etherscan.utils';
import { PriceData } from '../tokens/interfaces/price-data.interface';
import { getToken, getTokenByName } from '../tokens/tokens.utils';
import { CoinGeckoPriceResponse } from './interface/coingecko-price-response.interface';
import { TokenPrice } from './interface/token-price.interface';

export async function fetchPrices(chain: Chain, inputs: string[], lookupName = false): Promise<PriceData> {
  let baseURL;
  let params: Record<string, string>;

  // utilize coingecko name look up api
  if (lookupName) {
    baseURL = `${COINGECKO_URL}/price`;
    params = {
      ids: inputs.join(','),
      vs_currencies: 'usd',
    };
  } else {
    baseURL = `${COINGECKO_URL}/token_price/${chain.network}`;
    params = {
      contract_addresses: inputs.join(','),
      vs_currencies: 'usd',
    };
  }

  const expectedTokens: string[] = [];
  const result = await request<CoinGeckoPriceResponse>(baseURL, params);
  const priceData = Object.fromEntries(
    Object.entries(result).map((entry) => {
      const [key, value] = entry;
      let token;
      try {
        token = getToken(key);
      } catch {
        token = getTokenByName(key);
      }
      const { address } = token;
      expectedTokens.push(address);
      return [address, { address, price: value.usd }];
    }),
  );
  const missingTokens: string[] = [];
  expectedTokens.forEach((input) => {
    if (!priceData[input]) {
      missingTokens.push(input);
    }
  });
  if (missingTokens.length > 0) {
    console.error(`Missing prices for:\n${missingTokens.join('\n')}`);
  }
  return priceData;
}

/**
 * Retrieve the price data for a given token in USD and ETH.
 * Warning: Not all contracts are supported.
 * If a token is not supported, but available on CoinGecko, use getTokenPrice.
 * @param contract Token contract address.
 * @throws {InternalServerError} Failed price lookup.
 */
export async function getContractPrice(contract: string): Promise<TokenPrice> {
  const params = {
    contract_addresses: contract,
    vs_currencies: 'usd',
  };
  const result = await request<Record<string, { usd: number }>>(`${COINGECKO_URL}/token_price/ethereum`, params);
  const contractKey = contract.toLowerCase(); // coingecko return key in lower case
  if (!result[contractKey] || !result[contractKey].usd) {
    throw new InternalServerError(`Unable to resolve ${contract} price by contract`);
  }
  const token = getToken(contract);
  return {
    address: token.address,
    price: result[contractKey].usd,
  };
}

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param token CoinGecko token name.
 * @throws {InternalServerError} Failed price lookup.
 */
export async function getTokenPrice(name: string): Promise<TokenPrice> {
  const params = {
    ids: name,
    vs_currencies: 'usd',
  };
  const result = await request<Record<string, { usd: number }>>(`${COINGECKO_URL}/price`, params);
  if (!result[name] || !result[name].usd) {
    throw new InternalServerError(`Unable to resolve ${name} price by name`);
  }
  const token = getTokenByName(name);
  return {
    address: token.address,
    price: result[name].usd,
  };
}
