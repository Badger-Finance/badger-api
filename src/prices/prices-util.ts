import { BadRequest, InternalServerError, NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { DocumentClient, QueryInput } from 'aws-sdk/clients/dynamodb';
import { ethers } from 'ethers';
import NodeCache = require('node-cache');
import { getItems, saveItem } from '../aws/dynamodb-utils';
import { ChainStrategy } from '../chains/strategies/chain.strategy';
import { COINGECKO_URL, PRICE_DATA, TOKENS } from '../config/constants';
import { Token } from '../tokens/interfaces/token.interface';
import { PriceData, TokenPrice, TokenPriceSnapshot } from '../tokens/interfaces/token-price.interface';
import { getToken, getTokenByName, protocolTokens } from '../tokens/tokens-util';
import { TokenConfig } from '../tokens/types/token-config.type';
import AttributeValue = DocumentClient.AttributeValue;
import fetch from 'node-fetch';
import { loadChains } from '../chains/chain';
import { Chain } from '../chains/config/chain.config';
import { getSett } from '../setts/setts-util';
import { TokenType } from '../tokens/enums/token-type.enum';

/**
 * Protoype for a token address pricing function.
 * @param address Target for price retrieval.
 */
export type PricingFunction = (address: string) => Promise<TokenPrice>;

/**
 * Mass price update request object.
 * Mapping of token address to respective pricing function.
 * @see {PricingFunction}
 */
export interface PriceUpdateRequest {
  [token: string]: PricingFunction;
}

export const priceCache = new NodeCache({ stdTTL: 300, checkperiod: 480 });

/**
 * Update pricing db entry using chain strategy.
 * @param token Target for price update.
 */
export const updatePrice = async (token: Token): Promise<void> => {
  const { address, name } = token;
  const strategy = ChainStrategy.getStrategy(address);
  const tokenPriceData = await strategy.getPrice(address);
  tokenPriceData.name = name;
  tokenPriceData.address = address;
  const tokenPriceSnapshot: TokenPriceSnapshot = {
    ...tokenPriceData,
    updatedAt: Date.now(),
  };
  await saveItem(PRICE_DATA, tokenPriceSnapshot);
};

/**
 * Mass update token pricing db entries.
 * @param tokenConfig Target for price updates.
 */
export const updatePrices = async (tokenConfig: TokenConfig): Promise<void> => {
  await Promise.all(Object.values(tokenConfig).map(async (token) => updatePrice(token)));
};

/**
 * Load token price fromt he pricing database.
 * @param contract Address for the token price being requested.
 * @returns Most recent price data for the requested contract.
 */
export const getPrice = async (contract: string): Promise<TokenPriceSnapshot> => {
  const checksumContract: AttributeValue = ethers.utils.getAddress(contract);
  const params: QueryInput = {
    TableName: PRICE_DATA,
    KeyConditionExpression: 'address = :address',
    ExpressionAttributeValues: {
      ':address': checksumContract,
    },
    Limit: 1,
    ScanIndexForward: false,
  };
  const prices = await getItems<TokenPriceSnapshot>(params);
  if (!prices || prices.length !== 1) {
    throw new NotFound(`No price data stored for ${checksumContract}`);
  }
  return prices[0];
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param contract Token contract address.
 * @returns Most recently updated token pricing data.
 */
export const getTokenPriceData = async (contract: string): Promise<TokenPrice> => {
  const checksumContract = ethers.utils.getAddress(contract);
  const cachedPrice = priceCache.get<TokenPrice>(checksumContract);
  if (!cachedPrice) {
    const priceData = await getPriceData();
    return priceData[checksumContract];
  }
  return cachedPrice;
};

/**
 * Retrieve all chain token prices in both USD and ETH.
 * @returns Most recently updated token pricing data for all tokens.
 */
export const getPriceData = async (): Promise<PriceData> => {
  const priceData: PriceData = {};
  const prices = await Promise.all(Object.keys(protocolTokens).map((key) => getPrice(protocolTokens[key].address)));
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  prices.forEach((token) => {
    priceData[token.address] = token;
    priceCache.set(token.address, token);
  });
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  return priceData;
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * Warning: Not all contracts are supported.
 * If a token is not supported, but available on CoinGecko, use getTokenPrice.
 * @param contract Token contract address.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getContractPrice = async (contract: string): Promise<TokenPrice> => {
  const cachedPrice = priceCache.get<TokenPrice>(contract);
  if (cachedPrice) {
    return cachedPrice;
  }
  const response = await fetch(
    `${COINGECKO_URL}/token_price/ethereum?contract_addresses=${contract}&vs_currencies=usd,eth`,
  );
  if (!response.ok) {
    throw new InternalServerError(`Unable to query ${contract} price`);
  }
  const json = await response.json();
  const contractKey = contract.toLowerCase(); // coingecko return key in lower case
  if (!json[contractKey] || !json[contractKey].usd || !json[contractKey].eth) {
    throw new InternalServerError(`Unable to resolve ${contract} price`);
  }
  const token = getToken(contract);
  const contractPrice: TokenPrice = {
    name: token.name,
    address: token.address,
    usd: json[contractKey].usd,
    eth: json[contractKey].eth,
  };
  priceCache.set(contract, contractPrice);
  return contractPrice;
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param token CoinGecko token name.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getTokenPrice = async (name: string): Promise<TokenPrice> => {
  const cachedPrice = priceCache.get<TokenPrice>(name);
  if (cachedPrice) {
    return cachedPrice;
  }
  const response = await fetch(`${COINGECKO_URL}/price?ids=${name}&vs_currencies=usd,eth`);
  if (!response.ok) {
    throw new InternalServerError(`Unable to query ${name} price`);
  }
  const json = await response.json();
  if (!json[name] || !json[name].usd || !json[name].eth) {
    throw new InternalServerError(`Unable to resolve ${name} price`);
  }
  const token = getTokenByName(name);
  const tokenPrice: TokenPrice = {
    name: token.name,
    address: token.address,
    usd: json[name].usd,
    eth: json[name].eth,
  };
  priceCache.set(name, tokenPrice);
  return tokenPrice;
};

/**
 * Convert a given token price to a defined currency option.
 * @param tokenPrice Pricing data for the given token.
 * @param currency Currency requested from the pricing data.
 * @returns Price value in currency if exists, default to usd if not.
 */
export const inCurrency = (tokenPrice: TokenPrice, currency?: string): number => {
  switch (currency) {
    case 'eth':
      return tokenPrice.eth;
    case 'usd':
    default:
      return tokenPrice.usd;
  }
};

/**
 * Get pricing information for a vault token.
 * @param contract Address for vault token.
 * @returns Pricing data for the given vault token based on the pricePerFullShare.
 */
export const getVaultTokenPrice = async (contract: string): Promise<TokenPrice> => {
  const token = getToken(contract);
  if (token.type !== TokenType.Vault) {
    throw new BadRequest(`${token.name} is not a vault token`);
  }
  if (!token.vaultToken) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const vaultToken = token.vaultToken;
  const targetChain = Chain.getChain(vaultToken.network);
  const [vaultTokenPrice, vaultTokenSnapshot] = await Promise.all([
    getPrice(vaultToken.address),
    getSett(targetChain.graphUrl, token.address),
  ]);
  if (!vaultTokenSnapshot.sett) {
    throw new InternalServerError(`Failed to load ${contract} sett data`);
  }
  vaultTokenPrice.name = token.name;
  vaultTokenPrice.address = token.address;
  const sett = vaultTokenSnapshot.sett;

  let multiplier;
  if (token.address === TOKENS.BDIGG) {
    multiplier = (sett.balance / sett.totalSupply) * 1e9;
  } else {
    multiplier = sett.pricePerFullShare / 1e18;
  }
  vaultTokenPrice.usd *= multiplier;
  vaultTokenPrice.eth *= multiplier;

  return vaultTokenPrice;
};

/**
 * Get pricing information for a wrapper token.
 * Wrapper tokens are tokens on their non-native chain that represent another token
 * in the system. i.e. bDigg on Binance Smart Chain.
 * @param contract Address for wrapper token.
 * @returns Pricing data for the given wrapper token.
 */
export const getWrapperTokenPrice = async (contract: string): Promise<TokenPrice> => {
  const token = getToken(contract);
  if (token.type !== TokenType.Wrapper) {
    throw new BadRequest(`${token.name} is not a wrapper token`);
  }
  if (!token.vaultToken) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const vaultToken = token.vaultToken;
  const allSetts = [];
  for (const chain of loadChains()) {
    allSetts.push(...chain.setts);
  }
  const backingVault = allSetts.find((sett) => ethers.utils.getAddress(sett.depositToken) === vaultToken.address);
  if (!backingVault) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const price = await getPrice(backingVault.settToken);
  if (!price) {
    throw new InternalServerError(`Failed to load ${contract} sett data`);
  }
  return price;
};
