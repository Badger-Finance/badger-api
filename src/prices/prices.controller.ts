import { Currency, Network } from '@badger-dao/sdk';
import { Controller, Inject } from '@tsed/di';
import { UseCache } from '@tsed/platform-cache';
import { QueryParams } from '@tsed/platform-params';
import { ContentType, Description, Get, Hidden, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { PriceSnapshots } from '../tokens/interfaces/price-snapshots.interface';
import { PriceSummary } from '../tokens/interfaces/price-summary.interface';
import { PriceSummaryModel } from '../tokens/interfaces/price-summary-model.interface';
import { PricesService } from './prices.service';

@Controller('/prices')
export class PricesController {
  @Inject()
  pricesService!: PricesService;

  @Get('')
  @UseCache()
  @ContentType('json')
  @Summary('Get a summary of token prices related to the Badger Protocol')
  @Description('Return a map of checksum contract address to the currency value of the token')
  @Returns(200, PriceSummaryModel)
  async listPrices(
    @QueryParams('tokens') tokens?: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency
  ): Promise<PriceSummary> {
    return this.pricesService.getPriceSummary(
      tokens?.split(',') ?? Object.keys(Chain.getChain(chain).tokens),
      currency
    );
  }

  @Get('/snapshots')
  @Hidden()
  @UseCache()
  @ContentType('json')
  async getPriceSnapshots(
    @QueryParams('tokens') tokens: string,
    @QueryParams('timestamps') timestamps: string,
    @QueryParams('currency') currency?: Currency
  ): Promise<PriceSnapshots> {
    return this.pricesService.getPriceSnapshots(
      tokens.split(','),
      timestamps.split(',').map((t) => Number(t)),
      currency
    );
  }
}
