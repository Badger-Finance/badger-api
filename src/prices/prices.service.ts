import { Currency } from '@badger-dao/sdk';
import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { PriceSummary } from '../tokens/interfaces/price-summary.interface';
import { convert, getPrice } from './prices.utils';

/**
 * API price oracle service. Uses CoinGecko as a source of truth for most
 * tokens when possible, and TheGraph for AMM pairs when not available
 * via CG. Prices are cached for 5 minutes at a time, but may live up to 8.
 */
@Service()
export class PricesService {
  async getPriceSummary(chain: Chain, currency?: Currency): Promise<PriceSummary> {
    const prices = await Promise.all(Object.keys(chain.tokens).map(async (token) => getPrice(token)));
    const entries = await Promise.all(
      prices.map(async (tokenPrice) => {
        const convertedPrice = await convert(tokenPrice.price, currency);
        return [tokenPrice.address, convertedPrice];
      }),
    );
    return Object.fromEntries(entries);
  }
}
