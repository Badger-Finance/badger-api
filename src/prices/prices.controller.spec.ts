import { PlatformTest } from '@tsed/common';
import { BadRequest } from '@tsed/exceptions';
import SuperTest from 'supertest';
import { Server } from '../Server';
import { bscTokensConfig } from '../tokens/config/bsc-tokens.config';
import { ethTokensConfig } from '../tokens/config/eth-tokens.config';
import { PricesService } from './prices.service';

describe('PricesController', () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;
  let pricesService: PricesService;

  beforeEach(PlatformTest.bootstrap(Server));
  beforeEach(async () => {
    request = SuperTest(PlatformTest.callback());
    pricesService = PlatformTest.get<PricesService>(PricesService);
  });

  afterEach(PlatformTest.reset);

  const getPrice = (address: string): number => parseInt(address.slice(0, 3), 16);

  describe('GET /v2/prices', () => {
    describe('with no specified chain', () => {
      it('returns eth token config', async (done: jest.DoneCallback) => {
        const result = Object.fromEntries(Object.keys(ethTokensConfig).map((token) => [token, getPrice(token)]));
        jest.spyOn(pricesService, 'getPriceSummary').mockImplementationOnce(() => Promise.resolve(result));
        const { body } = await request.get('/v2/prices').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with a specified chain', () => {
      it('returns the specified token config for eth', async (done: jest.DoneCallback) => {
        const result = Object.fromEntries(Object.keys(ethTokensConfig).map((token) => [token, getPrice(token)]));
        jest.spyOn(pricesService, 'getPriceSummary').mockImplementationOnce(() => Promise.resolve(result));
        const { body } = await request.get('/v2/prices?chain=eth').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });

      it('returns the specified token config for bsc', async (done: jest.DoneCallback) => {
        const result = Object.fromEntries(Object.keys(bscTokensConfig).map((token) => [token, getPrice(token)]));
        jest.spyOn(pricesService, 'getPriceSummary').mockImplementationOnce(() => Promise.resolve(result));
        const { body } = await request.get('/v2/prices?chain=bsc').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with an invalid specified chain', () => {
      it('returns a 400', async (done: jest.DoneCallback) => {
        const { body } = await request.get('/v2/prices?chain=invalid').expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });
});
