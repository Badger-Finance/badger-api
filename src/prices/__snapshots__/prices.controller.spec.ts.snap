// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricesController GET /v2/prices with a specified chain returns the specified token config for bsc 1`] = `
Object {
  "0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82": 0,
  "0x1F7216fdB338247512Ec99715587bb97BBf96eae": 1,
  "0x5986D5c77c65e5801a5cAa4fAE80089f870A71dA": 5,
  "0x5A58609dA96469E9dEf3fE344bC39B00d18eb9A5": 5,
  "0x61EB789d75A95CAa3fF50ed7E47b96c132fEc082": 6,
  "0x7130d2A12B9BCbFAe4f2634d864A1Ee1Ce3Ead9c": 7,
  "0x753fbc5800a8C8e3Fb6DC6415810d627A387Dfc9": 7,
  "0x7561EEe90e24F3b348E1087A005F78B4c8453524": 7,
  "0x81d776C90c89B8d51E9497D58338933127e2fA80": 8,
  "0x857F91f735f4B03b19D2b5c6E476C73DB8241F55": 8,
  "0xa861Ba302674b08f7F2F24381b705870521DDfed": 10,
  "0xaf4B9C4b545D5324904bAa15e29796D2E2f90813": 10,
  "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c": 11,
}
`;

exports[`PricesController GET /v2/prices with a specified chain returns the specified token config for eth 1`] = `
Object {
  "0x0316EB71485b0Ab14103307bf65a021042c6d380": 0,
  "0x04c90C198b2eFF55716079bc06d7CCc4aa4d7512": 0,
  "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3": 0,
  "0x090185f2135308BaD17527004364eBcC2D37e5F6": 0,
  "0x110492b31c59716AC47337E616804E3E3AdC0b4a": 1,
  "0x15cBC4ac1e81c97667780fE6DAdeDd04a6EEB47B": 1,
  "0x17d8CBB6Bce8cEE970a4027d1198F6700A7a6c24": 1,
  "0x1862A18181346EBd9EdAf800804f89190DeF24a5": 1,
  "0x18d98D452072Ac2EB7b74ce3DB723374360539f1": 1,
  "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28": 1,
  "0x19E4d89e0cB807ea21B8CEF02df5eAA99A110dA5": 1,
  "0x20c36f062a31865bED8a5B1e512D9a1A20AA333A": 2,
  "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599": 2,
  "0x235c9e24D3FB2FAFd58a2E49D454Fdcd2DBf7FF1": 2,
  "0x26B8efa69603537AC8ab55768b6740b67664D518": 2,
  "0x27E98fC7d05f54E544d16F58C194C2D7ba71e3B5": 2,
  "0x2B5455aac8d64C14786c3a29858E43b5945819C0": 2,
  "0x2fE94ea3d5d4a175184081439753DE15AeF9d614": 2,
  "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 3,
  "0x3c9d6c1C73b31c837832c72E04D3152f051fc1A9": 3,
  "0x410e3E86ef427e30B9235497143881f717d93c2A": 4,
  "0x48c59199Da51B7E30Ea200a74Ea07974e62C4bA7": 4,
  "0x49849C98ae39Fff122806C06791Fa73784FB3675": 4,
  "0x4b92d19c11435614CD49Af1b589001b7c08cD4D5": 4,
  "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B": 4,
  "0x5228a22e72ccC52d415EcFd199F99D0665E7733b": 5,
  "0x53C8E199eb2Cb7c01543C137078a038937a68E40": 5,
  "0x55912D0Cf83B75c492E761932ABc4DB4a5CB1b17": 5,
  "0x599D92B453C010b1050d31C364f6ee17E819f193": 5,
  "0x5Dce29e92b1b939F8E8C60DcF15BDE82A85be4a9": 5,
  "0x5a6A4D54456819380173272A5E8E9B9904BdF41B": 5,
  "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7": 6,
  "0x64eda51d3Ad40D56b9dFc5554E06F94e1Dd786Fd": 6,
  "0x6B175474E89094C44Da98b954EedeAC495271d0F": 6,
  "0x6B3595068778DD592e39A122f4f5a5cF09C90fE2": 6,
  "0x6aF7377b5009d7d154F36FE9e235aE1DA27Aea22": 6,
  "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490": 6,
  "0x6dEf55d2e18486B9dDfaA075bc4e4EE0B28c1545": 6,
  "0x758A43EE2BFf8230eeb784879CdcFF4828F2544D": 7,
  "0x798D1bE841a82a273720CE31c822C61a67a601C3": 7,
  "0x7e7E112A68d8D2E221E11047a72fFC1065c38e1a": 7,
  "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68": 8,
  "0x853d955aCEf822Db058eb8505911ED77F175b99e": 8,
  "0x85Eee30c52B0b379b046Fb0F85F4f3Dc3009aFEC": 8,
  "0x8751D4196027d4e6DA63716fA7786B5174F04C15": 8,
  "0x8798249c2E607446EfB7Ad49eC89dD1865Ff4272": 8,
  "0x88128580ACdD9c04Ce47AFcE196875747bF2A9f6": 8,
  "0x89Ab32156e46F46D02ade3FEcbe5Fc4243B9AAeD": 8,
  "0x8a8FFec8f4A0C8c9585Da95D9D97e8Cd6de273DE": 8,
  "0x8c76970747afd5398e958bDfadA4cf0B9FcA16c4": 8,
  "0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa": 8,
  "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551": 9,
  "0x945Facb997494CC2570096c74b5F66A3507330a1": 9,
  "0x99D8a9C45b2ecA8864373A26D1459e3Dff1e17F3": 9,
  "0x99F39D495C6A5237f43602f3Ab5F49786E46c9B0": 9,
  "0x9BE89D2a4cd102D8Fecc6BF9dA793be995C22541": 9,
  "0x9a13867048e01c663ce8Ce2fE0cDAE69Ff9F35E3": 9,
  "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48": 10,
  "0xAf5A1DECfa95BAF63E0084a35c62592B774A2A87": 10,
  "0xBC19712FEB3a26080eBf6f2F7849b417FdD792CA": 11,
  "0xBE08Ef12e4a553666291E9fFC24fCCFd354F2Dd2": 11,
  "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2": 12,
  "0xC17078FDd324CC473F8175Dc5290fae5f2E84714": 12,
  "0xCEfF51756c56CeFFCA006cD410B03FFC46dd3a58": 12,
  "0xD533a949740bb3306d119CC777fa900bA034cd52": 13,
  "0xDE5331AC4B3630f94853Ff322B66407e0D6331E8": 13,
  "0xE86204c4eDDd2f70eE00EAd6805f917671F56c52": 14,
  "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D": 14,
  "0xFbdCA68601f835b27790D98bbb8eC7f05FDEaA9B": 15,
  "0xa0246c9032bC3A600820415aE600c6388619A14D": 10,
  "0xa3BeD4E1c75D00fa6f4E5E6922DB7261B5E9AcD2": 10,
  "0xaE96fF08771a109dc6650a1BdCa62F2d558E40af": 10,
  "0xb19059ebb43466C323583928285a49f558E572Fd": 11,
  "0xb9D076fDe463dbc9f915E5392F807315Bf940334": 11,
  "0xc4AD29ba4B3c580e6D59105FFf484999997675Ff": 12,
  "0xc4E15973E6fF2A35cC804c2CF9D2a1b817a8b40F": 12,
  "0xcA3d75aC011BF5aD07a98d02f18225F9bD9A6BDF": 12,
  "0xcD7989894bc033581532D2cd88Da5db0A4b12859": 12,
  "0xd04c48A53c111300aD41190D63681ed3dAd998eC": 13,
  "0xd632f22692FaC7611d2AA1C0D552930D43CAEd3B": 13,
  "0xdAC17F958D2ee523a2206206994597C13D831ec7": 13,
  "0xf349c0faA80fC1870306Ac093f75934078e28991": 15,
  "0xfE18be6b3Bd88A2D2A7f928d00292E7a9963CfC6": 15,
  "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 15,
}
`;

exports[`PricesController GET /v2/prices with an invalid specified chain returns a 400 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`PricesController GET /v2/prices with no specified chain returns eth token config 1`] = `
Object {
  "0x0316EB71485b0Ab14103307bf65a021042c6d380": 0,
  "0x04c90C198b2eFF55716079bc06d7CCc4aa4d7512": 0,
  "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3": 0,
  "0x090185f2135308BaD17527004364eBcC2D37e5F6": 0,
  "0x110492b31c59716AC47337E616804E3E3AdC0b4a": 1,
  "0x15cBC4ac1e81c97667780fE6DAdeDd04a6EEB47B": 1,
  "0x17d8CBB6Bce8cEE970a4027d1198F6700A7a6c24": 1,
  "0x1862A18181346EBd9EdAf800804f89190DeF24a5": 1,
  "0x18d98D452072Ac2EB7b74ce3DB723374360539f1": 1,
  "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28": 1,
  "0x19E4d89e0cB807ea21B8CEF02df5eAA99A110dA5": 1,
  "0x20c36f062a31865bED8a5B1e512D9a1A20AA333A": 2,
  "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599": 2,
  "0x235c9e24D3FB2FAFd58a2E49D454Fdcd2DBf7FF1": 2,
  "0x26B8efa69603537AC8ab55768b6740b67664D518": 2,
  "0x27E98fC7d05f54E544d16F58C194C2D7ba71e3B5": 2,
  "0x2B5455aac8d64C14786c3a29858E43b5945819C0": 2,
  "0x2fE94ea3d5d4a175184081439753DE15AeF9d614": 2,
  "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 3,
  "0x3c9d6c1C73b31c837832c72E04D3152f051fc1A9": 3,
  "0x410e3E86ef427e30B9235497143881f717d93c2A": 4,
  "0x48c59199Da51B7E30Ea200a74Ea07974e62C4bA7": 4,
  "0x49849C98ae39Fff122806C06791Fa73784FB3675": 4,
  "0x4b92d19c11435614CD49Af1b589001b7c08cD4D5": 4,
  "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B": 4,
  "0x5228a22e72ccC52d415EcFd199F99D0665E7733b": 5,
  "0x53C8E199eb2Cb7c01543C137078a038937a68E40": 5,
  "0x55912D0Cf83B75c492E761932ABc4DB4a5CB1b17": 5,
  "0x599D92B453C010b1050d31C364f6ee17E819f193": 5,
  "0x5Dce29e92b1b939F8E8C60DcF15BDE82A85be4a9": 5,
  "0x5a6A4D54456819380173272A5E8E9B9904BdF41B": 5,
  "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7": 6,
  "0x64eda51d3Ad40D56b9dFc5554E06F94e1Dd786Fd": 6,
  "0x6B175474E89094C44Da98b954EedeAC495271d0F": 6,
  "0x6B3595068778DD592e39A122f4f5a5cF09C90fE2": 6,
  "0x6aF7377b5009d7d154F36FE9e235aE1DA27Aea22": 6,
  "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490": 6,
  "0x6dEf55d2e18486B9dDfaA075bc4e4EE0B28c1545": 6,
  "0x758A43EE2BFf8230eeb784879CdcFF4828F2544D": 7,
  "0x798D1bE841a82a273720CE31c822C61a67a601C3": 7,
  "0x7e7E112A68d8D2E221E11047a72fFC1065c38e1a": 7,
  "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68": 8,
  "0x853d955aCEf822Db058eb8505911ED77F175b99e": 8,
  "0x85Eee30c52B0b379b046Fb0F85F4f3Dc3009aFEC": 8,
  "0x8751D4196027d4e6DA63716fA7786B5174F04C15": 8,
  "0x8798249c2E607446EfB7Ad49eC89dD1865Ff4272": 8,
  "0x88128580ACdD9c04Ce47AFcE196875747bF2A9f6": 8,
  "0x89Ab32156e46F46D02ade3FEcbe5Fc4243B9AAeD": 8,
  "0x8a8FFec8f4A0C8c9585Da95D9D97e8Cd6de273DE": 8,
  "0x8c76970747afd5398e958bDfadA4cf0B9FcA16c4": 8,
  "0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa": 8,
  "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551": 9,
  "0x945Facb997494CC2570096c74b5F66A3507330a1": 9,
  "0x99D8a9C45b2ecA8864373A26D1459e3Dff1e17F3": 9,
  "0x99F39D495C6A5237f43602f3Ab5F49786E46c9B0": 9,
  "0x9BE89D2a4cd102D8Fecc6BF9dA793be995C22541": 9,
  "0x9a13867048e01c663ce8Ce2fE0cDAE69Ff9F35E3": 9,
  "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48": 10,
  "0xAf5A1DECfa95BAF63E0084a35c62592B774A2A87": 10,
  "0xBC19712FEB3a26080eBf6f2F7849b417FdD792CA": 11,
  "0xBE08Ef12e4a553666291E9fFC24fCCFd354F2Dd2": 11,
  "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2": 12,
  "0xC17078FDd324CC473F8175Dc5290fae5f2E84714": 12,
  "0xCEfF51756c56CeFFCA006cD410B03FFC46dd3a58": 12,
  "0xD533a949740bb3306d119CC777fa900bA034cd52": 13,
  "0xDE5331AC4B3630f94853Ff322B66407e0D6331E8": 13,
  "0xE86204c4eDDd2f70eE00EAd6805f917671F56c52": 14,
  "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D": 14,
  "0xFbdCA68601f835b27790D98bbb8eC7f05FDEaA9B": 15,
  "0xa0246c9032bC3A600820415aE600c6388619A14D": 10,
  "0xa3BeD4E1c75D00fa6f4E5E6922DB7261B5E9AcD2": 10,
  "0xaE96fF08771a109dc6650a1BdCa62F2d558E40af": 10,
  "0xb19059ebb43466C323583928285a49f558E572Fd": 11,
  "0xb9D076fDe463dbc9f915E5392F807315Bf940334": 11,
  "0xc4AD29ba4B3c580e6D59105FFf484999997675Ff": 12,
  "0xc4E15973E6fF2A35cC804c2CF9D2a1b817a8b40F": 12,
  "0xcA3d75aC011BF5aD07a98d02f18225F9bD9A6BDF": 12,
  "0xcD7989894bc033581532D2cd88Da5db0A4b12859": 12,
  "0xd04c48A53c111300aD41190D63681ed3dAd998eC": 13,
  "0xd632f22692FaC7611d2AA1C0D552930D43CAEd3B": 13,
  "0xdAC17F958D2ee523a2206206994597C13D831ec7": 13,
  "0xf349c0faA80fC1870306Ac093f75934078e28991": 15,
  "0xfE18be6b3Bd88A2D2A7f928d00292E7a9963CfC6": 15,
  "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 15,
}
`;
